tensorflow_addons/seq2seq/tests/decoder_test.py

```diff
@@ -15,169 +15,127 @@
         )
-
-    def testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNNNoSeqLen(self):
-        self._testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNN(
-            use_sequence_length=False
+        assert _t((batch_size, time_steps)) == final_outputs.sample_id.shape
+    np.testing.assert_array_equal(expected_length, final_sequence_length)
+
+
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+@pytest.mark.parametrize("maximum_iterations", [None, 0, 1])
+@pytest.mark.parametrize("time_major", [True, False])
+def test_dynamic_decode_rnn(time_major, maximum_iterations):
+    _test_decode_rnn(time_major, maximum_iterations)
+
+
+def _test_dynamic_decode_rnn_with_training_helper_matches_dynamic_rnn(
+    use_sequence_length,
+):
+    sequence_length = [3, 4, 3, 1, 0]
+    batch_size = 5
+    max_time = 8
+    input_depth = 7
+    cell_depth = 10
+    max_out = max(sequence_length)
+
+    inputs = np.random.randn(batch_size, max_time, input_depth).astype(np.float32)
+    inputs = tf.constant(inputs)
+
+    cell = tf.keras.layers.LSTMCell(cell_depth)
+    zero_state = cell.get_initial_state(batch_size=batch_size, dtype=tf.float32)
+    sampler = sampler_py.TrainingSampler()
+    my_decoder = basic_decoder.BasicDecoder(
+        cell=cell, sampler=sampler, impute_finished=use_sequence_length
+    )
+
+    (final_decoder_outputs, final_decoder_state, _,) = my_decoder(
+        inputs, initial_state=zero_state, sequence_length=sequence_length
+    )
+
+    rnn = tf.keras.layers.RNN(cell, return_sequences=True, return_state=True)
+    mask = (
+        tf.sequence_mask(sequence_length, maxlen=max_time)
+        if use_sequence_length
+        else None
+    )
+    outputs = rnn(inputs, mask=mask, initial_state=zero_state)
+    final_rnn_outputs = outputs[0]
+    final_rnn_state = outputs[1:]
+    if use_sequence_length:
+        final_rnn_outputs *= tf.cast(tf.expand_dims(mask, -1), final_rnn_outputs.dtype)
+
+    eval_result = {
+        "final_decoder_outputs": final_decoder_outputs,
+        "final_decoder_state": final_decoder_state,
+        "final_rnn_outputs": final_rnn_outputs,
+        "final_rnn_state": final_rnn_state,
+    }
+
+    # Decoder only runs out to max_out; ensure values are identical
+    # to dynamic_rnn, which also zeros out outputs and passes along
+    # state.
+    np.testing.assert_allclose(
+        eval_result["final_decoder_outputs"].rnn_output,
+        eval_result["final_rnn_outputs"][:, 0:max_out, :],
+    )
+    if use_sequence_length:
+        np.testing.assert_allclose(
+            eval_result["final_decoder_state"], eval_result["final_rnn_state"]
         )
+
+
+@pytest.mark.parametrize("use_sequence_length", [True, False])
+def test_dynamic_decode_rnn_with_training_helper_matches_dynamic_rnn_seq_len(
+    use_sequence_length,
+):
+    _test_dynamic_decode_rnn_with_training_helper_matches_dynamic_rnn(
+        use_sequence_length
+    )
```

Comment:
Same here, we can have only one function since we're using pytest.mark.parametrize.